src/User.php

```diff
@@ -70,7 +72,7 @@ public function getUnitId(): ?int
     /**
      * Vrací datum a čas automatického odhlášení ze skautISu
      */
-    public function getLogoutDate(): ?DateTime
+    public function getLogoutDate(): ?DateTimeInterface
```

Comment:
@JindrichPilar Nebylo by lepší tady používat přímo `DateTimeImmutable`?

Reply:
Jakou výhodu v tom vidíš?
Je to součást public interface a `updateLoginData` to bere jako argument. Preferoval bych nevázat se na konkrétní implementaci, i když v tuhle chvíli nejde implementovat DateTimeInterface ve vlastní třídě.

Reply:
Vesměs dva spolu úzce související důvody:

`DateTimeInterface` není obecně moc dobrej interface (je to leakující abstrakce) - vesměs kvůli mutatorům jako `add()`, `modify()` apod - pokud se chceš vyhnout bugům, musíš ideálně klonovat. Uvedu přímo na této metodě:

Metoda `getLogoutDate()` vrací  `DateTimeInterface`, tzn. buď `DateTimeImmutable`, nebo `DateTime`.

Pokud vrací `DateTime`, tak např. u volání:
```php
$minutaPredOdhlasenim = $user->getLogoutDate()->modify('- 1 minute');
```
nemůžu vědět, jestli jsem tím nezměnil logout date uvnitř uživatele (v tomto případě změnil).

Jedinné safe řešení tedy je
```
$minutaPredOdhlasenim = (clone $user->getLogoutDate())->modify('- 1 minute');
```

Navíc pokud chci předávat kamkoliv dál, kde používám DateTimeImmutable (protože se této nejistotě a klonování na hranici každého objektu ve svém kódu vyhýbám) musím udělat následující:

```
$logoutDate = $user->getLogoutDate();

doSomethingWithLogoutDate($logoutDate instanceof DateTime ? DateTimeImmutable::createFromMutable($logoutDate) : $logoutDate);
``` 

Oproti tomu při vracení `DateTimeImmutable` můžů volat na té instanci bezpečně cokoliv. Navíc `DateTime` z toho jednoduše dostanu (na PHP 7.3+) pomocí:
```php
DateTime::createFromImmutable($user->getLogoutDate());
```

Reply:
Ještě částečně chápu použití `DateTimeInterface` pro argumenty `setLogoutDate(DateTimeInterface $logoutDate) : void`, i když tam pak uvnitř je potřeba klonovat - rozšíříš tím (IMO stejně zbytečně) použitelnost API. Oproti tomu u return typu je to jako kdybys vracel `DateTime|DateTimeImmutable` - znamená to pro značnou část použití nutnost ošetřit obě varianty

Reply:
> vesměs kvůli mutatorům jako add(), modify()

> Oproti tomu u return typu je to jako kdybys vracel DateTime|DateTimeImmutable - znamená to pro značnou část použití nutnost ošetřit obě varianty

> i když tam pak uvnitř je potřeba klonovat

Ne. `DateTimeInterface` nedefinuje `modify` ani jiné podobné metody. Volat to na tom je prasárna. 

Ale ok, immutable je snazší na použití a nezkušený člověk má menší šanci napsat nesmysl.

Reply:
Hustý, jsem ani nevěděl, že tyto metody definované nemá, díky za info, jsem na něj koukal v době kdy byl přidán, a očividně si to pamatuju špatně. V tom případě by jejich volání každopádně bylo prasárna.

Každopádně tím spíš mi ten interface přijde úplně k ničemu :) a zbytek s neergonomickým používáním tedy IMO stále platí.